File: src/car.cpp
#include "car.hpp"


namespace robot
{
    
bool Car::setup(ros::NodeHandle nh, ros::NodeHandle nh_priv)
{
    nav_finish_flag = false;
    init_pose = geometry_msgs::Pose2D();
    pose = geometry_msgs::Pose2D();
    velocity = geometry_msgs::Twist();

    // Parameter
    vector<double> _polygon;
    nh_priv.param<double>("duration", duration, 0.1);
    nh_priv.param<string>("map_frame", map_frame, "map");
    nh_priv.param<string>("odom_frame", odom_frame, "odom");
    nh_priv.param<string>("base_frame", base_frame, "base_link");
    nh_priv.param<double>("x", init_pose.x, 0);      // Initial pose of car
    nh_priv.param<double>("y", init_pose.y, 0);
    nh_priv.param<double>("theta", pose.theta, 0);
    nh_priv.param<double>("vx", velocity.linear.x, 0); // Initial velocity of car
    nh_priv.param<double>("vy", velocity.linear.y, 0);
    nh_priv.param<double>("vth", velocity.angular.z, 0);
    nh_priv.getParam("max_vel", max_vel);
    nh_priv.getParam("polygon", _polygon);

    pose.theta = DEG2RAD(pose.theta);
    for(int i=0;i<_polygon.size() / 2;i++)
    {
        geometry_msgs::Point32 point;
        point.x = _polygon[2 * i];
        point.y = _polygon[2 * i + 1];
        polygon.points.push_back(point);
    }

    // Subscribe velocity topic
    vel_sub = nh.subscribe("cmd_vel", 10, &Car::VelCallback, this);
    goalreached_sub = nh.subscribe("isgoalreached", 10, &Car::GoalReachedCallback, this);

    // Advertise odometry topic
    odom_pub = nh_priv.advertise<nav_msgs::Odometry>("odom", 10);
    path_pub = nh_priv.advertise<nav_msgs::Path>("path", 10);
    polygon_pub = nh_priv.advertise<geometry_msgs::PolygonStamped>("polygon", 10);

    path = nav_msgs::Path();
    path.header.stamp = ros::Time::now();
    path.header.frame_id = odom_frame;

    // Publish static TF
    PublishStaticTF();

    return true;
}

void Car::spin()
{
    while(ros::ok())
    {
        update();
        PublishOdometry();
        PublishPath();
        PublishTF();
        PublishPolygon();

        // Sleep and wait for callback
        ros::Duration(duration).sleep();
        ros::spinOnce();
    }    
}

void Car::VelCallback(const geometry_msgs::Twist& msg)
{
    if(nav_finish_flag)
    {
        velocity = geometry_msgs::Twist();
    }
    else
    {
        velocity.linear.x = abs(msg.linear.x) > max_vel[0] ? (msg.linear.x > 0 ? max_vel[0] : -max_vel[0]) : msg.linear.x;
        velocity.linear.y = abs(msg.linear.y) > max_vel[1] ? (msg.linear.y > 0 ? max_vel[1] : -max_vel[1]) : msg.linear.y;
        velocity.angular.z = abs(msg.angular.x) > max_vel[2] ? (msg.angular.x > 0 ? max_vel[2] : -max_vel[2]) : msg.angular.x;

        // ROS_INFO("velocity: %f, %f, %f", velocity.linear.x, velocity.linear.y, velocity.linear.z);
    }
}

void Car::GoalReachedCallback(const std_msgs::Bool& msg)
{
    if(!nav_finish_flag && msg.data)
        ROS_INFO("Finishing navigation");
    nav_finish_flag = msg.data;
}

void Car::update()
{
    // Calculate the pose of the car
    double vx = velocity.linear.x;
    double vy = velocity.linear.y;
    double vth = velocity.angular.z;

    double dt = duration;

    double dx = (vx * cos(pose.theta) - vy * sin(pose.theta)) * dt;
    double dy = (vx * sin(pose.theta) + vy * cos(pose.theta)) * dt;
    double dth = vth * dt;

    pose.x += dx;
    pose.y += dy;
    pose.theta += dth;
    pose.theta = fmod(pose.theta, 2 * C_PI);    
}

void Car::PublishOdometry()
{
    // Publish odometry topic
    nav_msgs::Odometry odom;
    odom.header.stamp = ros::Time::now();
    odom.header.frame_id = odom_frame;

    odom.pose.pose.position.x = pose.x;
    odom.pose.pose.position.y = pose.y;
    odom.pose.pose.position.z = 0.0;
    odom.pose.pose.orientation = tf::createQuaternionMsgFromYaw(pose.theta);

    odom.child_frame_id = base_frame;
    odom.twist.twist.linear.x = velocity.linear.x;
    odom.twist.twist.linear.y = velocity.linear.y;
    odom.twist.twist.linear.z = 0.0;    
    odom.twist.twist.angular.z = velocity.angular.z;

    odom_pub.publish(odom);
}

void Car::PublishPath()
{
    // Publish path topic
    geometry_msgs::PoseStamped pose_stamped = geometry_msgs::PoseStamped();
    pose_stamped.header.stamp = ros::Time::now();
    pose_stamped.header.frame_id = odom_frame;
    pose_stamped.pose.position.x = pose.x;
    pose_stamped.pose.position.y = pose.y;
    pose_stamped.pose.position.z = 0.0;      
    path.poses.push_back(pose_stamped);
    path_pub.publish(path);
}

void Car::PublishTF()
{
    static tf2_ros::TransformBroadcaster broadcaster;
    geometry_msgs::TransformStamped transformStamped;

    transformStamped.header.stamp = ros::Time::now();
    transformStamped.header.frame_id = odom_frame;
    transformStamped.child_frame_id = base_frame;
    transformStamped.transform.translation.x = pose.x;
    transformStamped.transform.translation.y = pose.y;
    transformStamped.transform.translation.z = 0.0;
    transformStamped.transform.rotation = tf::createQuaternionMsgFromYaw(pose.theta);
    broadcaster.sendTransform(transformStamped);
}

void Car::PublishStaticTF()
{
    static tf2_ros::StaticTransformBroadcaster static_broadcaster;
    geometry_msgs::TransformStamped static_transformStamped;

    static_transformStamped.header.stamp = ros::Time::now();
    static_transformStamped.header.frame_id = map_frame;
    static_transformStamped.child_frame_id = odom_frame;
    static_transformStamped.transform.translation.x = init_pose.x;
    static_transformStamped.transform.translation.y = init_pose.y;
    static_transformStamped.transform.translation.z = 0.0;
    static_transformStamped.transform.rotation = tf::createQuaternionMsgFromYaw(init_pose.theta);
    static_broadcaster.sendTransform(static_transformStamped);
}

void Car::PublishPolygon()
{
    geometry_msgs::PolygonStamped polygonStamped;
    polygonStamped.header.stamp = ros::Time::now();
    polygonStamped.header.frame_id = base_frame;
    polygonStamped.polygon = polygon;
    polygon_pub.publish(polygonStamped);
}

} 

File: python/car.py
import numpy as np
import rospy
from geometry_msgs.msg import Pose2D
from geometry_msgs.msg import Twist
from geometry_msgs.msg import Polygon
from geometry_msgs.msg import PolygonStamped
from geometry_msgs.msg import Point32
from geometry_msgs.msg import PoseStamped
from geometry_msgs.msg import TransformStamped
from nav_msgs.msg import Odometry
from nav_msgs.msg import Path
from std_msgs.msg import Bool
import tf2_ros
import tf

deg2rad = lambda x: x * np.pi / 180.0
rad2deg = lambda x: x * 180.0 / np.pi

class Car:
    def __init__(self):
        self.nav_finish_flag = False
        self.init_pose = Pose2D()
        self.pose = Pose2D()
        self.velocity = Twist()
        self.polygon = Polygon()

    def setup(self):

        # Parameter
        self.duration = rospy.get_param('~duration', 0.1)
        self.map_frame = rospy.get_param('~map_frame', 'map')
        self.odom_frame = rospy.get_param('~odom_frame', 'odom')
        self.base_frame = rospy.get_param('~base_frame', 'base_link')
        self.init_pose.x = rospy.get_param('~x', 0.0)  # Initial pose of drone
        self.init_pose.y = rospy.get_param('~y', 0.0)
        self.pose.theta = rospy.get_param('~theta', 0.0) 
        self.velocity.linear.x = rospy.get_param('~vx', 0.0)   # Initial velocity of drone
        self.velocity.linear.y = rospy.get_param('~vy', 0.0)
        self.velocity.angular.z = rospy.get_param('~vyaw', 0.0)
        self.max_vel = rospy.get_param('~max_vel', [1.0, 1.0, 1.0])
        polygon = rospy.get_param('~polygon')
        
        self.pose.theta = deg2rad(self.pose.theta)
        for i in range(len(polygon) // 2):
            point = Point32()
            point.x = polygon[2 * i]
            point.y = polygon[2 * i + 1]   
            self.polygon.points.append(point)

        # Subscribe velocity topic
        rospy.Subscriber('cmd_vel', Twist, self.VelCallback, queue_size=10)
        rospy.Subscriber('isgoalreached', Bool, self.GoalReachedCallback, queue_size=10)

        # Advertise odometry topic
        self.odom_pub = rospy.Publisher('~odom', Odometry, queue_size=10)
        self.path_pub = rospy.Publisher('~path', Path, queue_size=10)
        self.polygon_pub = rospy.Publisher('~polygon', PolygonStamped, queue_size=10)

        self.path = Path()
        self.path.header.stamp = rospy.Time().now()
        self.path.header.frame_id = self.odom_frame

        # Publish static TF
        self.PublishStaticTF()

        return True
    
    def spin(self):
        while not rospy.is_shutdown():
            self.update()
            self.PublishOdometry()
            self.PublishPath()
            self.PublishTF()
            self.PublishPolygon()
            
            # Sleep and wait for callback
            rospy.sleep(self.duration)

    def VelCallback(self, msg):
        if self.nav_finish_flag:
            self.velocity = Twist()
        else:
            self.velocity.linear.x = float(np.clip(msg.linear.x, -self.max_vel[0], self.max_vel[0]))
            self.velocity.linear.y = float(np.clip(msg.linear.y, -self.max_vel[1], self.max_vel[1]))
            self.velocity.angular.z = float(np.clip(msg.angular.z, -self.max_vel[2], self.max_vel[2]))

            # rospy.loginfo('velocity: %f, %f, %f', self.velocity.linear.x, self.velocity.linear.y, self.velocity.angular.z)

    def GoalReachedCallback(self, msg):
        if not self.nav_finish_flag and msg.data:
            rospy.loginfo('Finishing navigation')
        self.nav_finish_flag = msg.data

    def update(self):
        # Calculate the pose of the drone
        vx = self.velocity.linear.x
        vy = self.velocity.linear.y
        vth = self.velocity.angular.z
        theta = self.pose.theta
        dt = self.duration

        dx = (vx * np.cos(theta) - vy * np.sin(theta)) * dt
        dy = (vx * np.sin(theta) + vy * np.cos(theta)) * dt
        dth = vth * dt

        self.pose.x += dx
        self.pose.y += dy
        self.pose.theta = (theta + dth) % (2 * np.pi)

    def PublishOdometry(self):
        quat = tf.transformations.quaternion_from_euler(0.0, 0.0, self.pose.theta)

        # Publish odometry topic
        odom = Odometry()
        odom.header.stamp = rospy.Time().now()
        odom.header.frame_id = self.odom_frame

        odom.pose.pose.position.x = self.pose.x
        odom.pose.pose.position.y = self.pose.y
        odom.pose.pose.position.z = 0.0
        odom.pose.pose.orientation.x = quat[0]
        odom.pose.pose.orientation.y = quat[1]
        odom.pose.pose.orientation.z = quat[2]
        odom.pose.pose.orientation.w = quat[3]
        odom.twist.twist = self.velocity
        self.odom_pub.publish(odom)

    def PublishPath(self):
        # Publish path topic
        pose_stamped = PoseStamped()
        pose_stamped.header.stamp = rospy.Time().now()
        pose_stamped.header.frame_id = self.odom_frame
        pose_stamped.pose.position.x = self.pose.x
        pose_stamped.pose.position.y = self.pose.y
        pose_stamped.pose.position.z = 0.0
        self.path.poses.append(pose_stamped)
        self.path_pub.publish(self.path)

    def PublishTF(self):
        quat = tf.transformations.quaternion_from_euler(0.0, 0.0, self.pose.theta)

        # Publish tf
        broadcaster = tf2_ros.TransformBroadcaster()
        transformStamped = TransformStamped()
        transformStamped.header.stamp = rospy.Time().now()
        transformStamped.header.frame_id = self.odom_frame
        transformStamped.child_frame_id = self.base_frame
        transformStamped.transform.translation.x = self.pose.x
        transformStamped.transform.translation.y = self.pose.y
        transformStamped.transform.translation.z = 0.0
        transformStamped.transform.rotation.x = quat[0]
        transformStamped.transform.rotation.y = quat[1]
        transformStamped.transform.rotation.z = quat[2]
        transformStamped.transform.rotation.w = quat[3]
        broadcaster.sendTransform(transformStamped)

    def PublishStaticTF(self):
        static_broadcaster = tf2_ros.StaticTransformBroadcaster()
        static_transformStamped = TransformStamped()
        static_transformStamped.header.stamp = rospy.Time().now()
        static_transformStamped.header.frame_id = self.map_frame
        static_transformStamped.child_frame_id = self.odom_frame
        static_transformStamped.transform.translation.x = self.init_pose.x
        static_transformStamped.transform.translation.y = self.init_pose.y
        static_transformStamped.transform.translation.z = 0.0
        static_transformStamped.transform.rotation.x = 0
        static_transformStamped.transform.rotation.y = 0
        static_transformStamped.transform.rotation.z = 0
        static_transformStamped.transform.rotation.w = 1
        static_broadcaster.sendTransform(static_transformStamped)

    def PublishPolygon(self):
        polygonStamped = PolygonStamped()
        polygonStamped.header.stamp = rospy.Time().now()
        polygonStamped.header.frame_id = self.base_frame
        polygonStamped.polygon = self.polygon
        self.polygon_pub.publish(polygonStamped)